src/SFA.DAS.EAS.Portal.Worker/EventHandlers/Commitments/CohortApprovalRequestedByProviderEventHandler.cs

```diff
@@ -0,0 +1,44 @@
+﻿using NServiceBus;
+using SFA.DAS.CommitmentsV2.Messages.Events;
+using SFA.DAS.EAS.Portal.Application.Adapters;
+using SFA.DAS.EAS.Portal.Application.Commands;
+using SFA.DAS.EAS.Portal.Application.Commands.Cohort;
+using SFA.DAS.EAS.Portal.Application.Services;
+using System;
+using System.Globalization;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EAS.Portal.Worker.EventHandlers.Commitments
+{  
+    public class CohortApprovalRequestedByProviderEventHandler : IHandleMessages<CohortApprovalRequestedByProvider>
+    {
+        private readonly IMessageContext _messageContext;
+        private readonly ICommandHandler<CohortApprovalRequestedCommand> _handler;
+        private readonly IAdapter<CohortApprovalRequestedByProvider, CohortApprovalRequestedCommand> _adapter;
+
+        const string TimeSentFormat = "yyyy-MM-dd HH:mm:ss:ffffff Z";
+
+        public CohortApprovalRequestedByProviderEventHandler(
+            ICommandHandler<CohortApprovalRequestedCommand> handler,
+            IAdapter<CohortApprovalRequestedByProvider, CohortApprovalRequestedCommand> adapter,
+            IMessageContext messageContext)
```

Comment:
`CohortApprovalRequestedByProviderEventHandler` executes in the context of the hosted service we create to host the NServiceBus endpoint.

From Microsoft's [docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-2.2), 'No scope is created for a hosted service by default.', and there is no manual creation of a scope using `CreateScope()`.

The injected `IMessageContext` is added to the ServiceCollection with Scoped lifetime. What then is the lifetime behaviour of `IMessageContext`?

I suspect this is going to run into multi-threaded issues. Can you reassure me that there will be no multi-threaded issues? Has this been tested processing many messages concurrently?

Reply:
Will investigate further but I haven't seen any evidence why there would be multi-threaded issues.

Reply:
I've looked into this and the DI for the event handlers is provided by NService bus not the hosted webjob.  In NServiceBus (because we are using the ServicesBuilder) scoped Microsoft dependencies are mapped to InstancePerUnitOfWork NServiceBus dependencies which (unless coded to behave differently) equates to the processing  of a single transport message.  Therefore this use case will work as expected.